packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -147,33 +225,109 @@ contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
     internal
     override
     {
-        (address payer,) = abi.decode(context, (address, uint256));
-        uint256 ethActualCharge = relayHub.calculateCharge(gasUseWithoutPost + gasUsedByPost, relayData);
-        _depositProceedsToHub(ethActualCharge);
-        uint256 remainingTokenBalance = token.balanceOf(address(this));
-        require(token.transfer(payer, remainingTokenBalance), "failed refund");
-        emit TokensCharged(gasUseWithoutPost, gasUsedByPost, ethActualCharge);
-    }
-
-    function _depositProceedsToHub(uint256 ethActualCharge) private {
-        UniswapV3Helper.swapToEth(
-            address(weth),
-            address(token),
-            ethActualCharge,
-            uniswapPoolFee,
-            uniswap
-        );
-        relayHub.depositFor{value : address(this).balance}(address(this));
+        (address payer, uint256 priceQuote, uint256 tokenPreCharge) = abi.decode(context, (address, uint256, uint256));
+        IERC20Metadata token = _getTokenFromPaymasterData(relayData.paymasterData);
+        (uint256 tokenActualCharge, uint256 ethActualCharge) = _calculateCharge(relayData, gasUseWithoutPost + gasUsedByPost, priceQuote, tokensSwapData[token].priceDivisor);
+        require(tokenActualCharge <= tokenPreCharge, "actual charge higher");
+        token.safeTransfer(payer, tokenPreCharge - tokenActualCharge);
+
+        emit TokensCharged(gasUseWithoutPost, gasUsedByPost, tokenActualCharge, ethActualCharge);
+        _refillHubDepositIfNeeded(ethActualCharge);
+        _withdrawToOwnerIfNeeded();
+    }
+
+    function _refillHubDepositIfNeeded(uint256 ethActualCharge) private {
+        uint256 hubBalance = relayHub.balanceOf(address(this));
+        if (hubBalance >= minHubBalance + ethActualCharge) {
+            return;
+        }
+        uint256 depositAmount = targetHubBalance - hubBalance + ethActualCharge;
+        _refillHubDeposit(depositAmount);
+    }
+
+    function _refillHubDeposit(uint256 depositAmount) private {
+        uint256 balance = address(this).balance;
+        uint256 amountSwapped = 0;
+        if (balance < depositAmount) {
+            for (uint256 i = 0; i < tokens.length && balance + amountSwapped < depositAmount; i++) {
+                IERC20Metadata tokenIn = tokens[i];
+                uint256 tokenBalance = tokenIn.balanceOf(address(this));
+                if (tokenBalance > 0) {
```

Comment:
0 may be a bar way too low. I think it would actually be cheaper to sort tokens by total value or something like that, maybe as an alternative path, or we can get a hint on what to do as another parameter in `paymasterData`. Here you would always start swapping with the same token so it is likely you will be going the opposite way - going from the least valuable to the most valuable one.

Reply:
Added a threshold